### ISSUE TITLE
Constructor arguments swapping if one (not the first) of them is NULL

### ISSUE THREAD

**Issue by KerbalScientist** [TomBZombie/Dice#62]
Branch: PHP5.4---PHP5.5
I believe the master branch is also affected, but didn't test it.

Code:
```php
class A {
  function __construct($must_be_string, $may_be_null)
  {
    var_export(func_get_args());
  }
}
$dice = new \Dice\Dice();
$dice->create('A', ['string', NULL]);
```

Output:
```php
array (
  0 => NULL,
  1 => 'string',
)
```
Expected output:
```php
array (
  0 => 'string',
  1 => NULL,
)
```

Plausible cause:
I think the problem is on line 74 of dice.php:
```php
if ($class && $args[$i] instanceof $class || !$args[$i] && $allowsNull) {
    $parameters[] = array_splice($args, $i, 1)[0];
    continue 2;
}
```
`!$args[$i] && $allowsNull` is evaluated to TRUE when `$i` is 1 and `foreach` loop above handles the first item of `$paramInfo` (so `$parameters` is empty) and the next line splices the second NULL argument from `$args`.
To be honest, I failed to fully understand, why `!$args[$i] && $allowsNull` is needed. I think it may be not needed at all, because NULL arguments should be correctly handled by `$this->expand(array_shift($args))` below.

If i am right, it may be sort of code style related issue. When you are trying to place too much logic in one line without documenting it, you start to forget what was your initial intention when you wrote it. So, you can fail to review it properly. No offends, maybe I'm just not smart enough to freely read code like this.

**Comment by gcaplan** [comment]
I'm experiencing this too, but with an empty array in the argument rather than a NULL. I'm using the last PHP 5.5 release, as I don't have access to 5.6 yet.

I'd really like to use Dice - it's a brilliant design. But this is a show-stopper for me, and as the code is, how can I put this, a little compressed and tough to read :-) I suspect it would be easier for you to find the fix.

Are you around? Can you tidy this up for us? I'm on a tight schedule...

**Comment by daniel-meister** [comment]
See my proposed pull request to fix those issues; do the test cases provided in commit d21818622d379b59bb3db7a17f67d9bc0e22d8b2 accurately describe your issues?

**Comment by gcaplan** [comment]
Hi daniel-meister

Yes - testEmptyArraySwap() replicates my issue. Many thanks for the patch!

**Comment by TomBZombie** [comment]
Thanks, I've fixed this issue in the PHP5.4-5.5 branch. It was resolved in master but not backported.

**Comment by KerbalScientist** [comment]
Unfortunately, this change does't fix the issue. Also, you may want to recheck the master branch, I'm shure it's also affected.
I feel like I'm supposed to propose a solution, but it can be not so simple. I think we first need to clearly describe the logic of argument handling and we may find out that trying to allow user to override parameters, provided by rules by arguments, given to Dice::create is not so good idea, while we are trying to map arguments by type-hinting, not by names as Aura.Di does. May be, we just don't have anough information to handle arguments unambiguously this way. 
I was on schedule, so I just switched to Aura.Di. Sorry, but I can't give much time to this issue.

**Comment by TomBZombie** [comment]
Thanks for pointing this out, it was a slightly different issue that was fixed above. I have found the source of the problem. As you mentioned:

> To be honest, I failed to fully understand, why !$args[$i] && $allowsNull is needed. I 

This is exactly the source of the problem. This check is required to allow `null` to be passed instead of a class. Without this, Dice will construct an instance of the hinted class when it sees `__construct(A $a = null)`, the allowsnull check ensures Dice passes null when instructed to explicitly.

The problem is this logic as you discovered. This check should only be run when a class name is type hinted.

This is solved by only running the check when a class is type hinted:

```
if ($class && $args[$i] instanceof $class || !$args[$i] && $allowsNull) {
```

Becomes

```
if ($class && ($args[$i] instanceof $class || !$args[$i] && $allowsNull)) {
```
